core/src/main/java/com/pipai/wf/battle/action/PrecisionShotAction.java

```diff
@@ -49,6 +51,10 @@ public PercentageModifierList getCritCalculation() {
 	@Override
 	protected void performImpl() throws IllegalActionException {
 		Agent a = getPerformer();
+		Ability ability = a.getInnateAbilities().getAbility(PrecisionShotAbility.class);
+		if (ability == null) {
+			ability = a.getWeaponGrantedAbilities().getAbility(PrecisionShotAbility.class);
+		}

```

Comment:
Can we change this so that we can just call a.getAbility(x)? It will probably be used pretty often.

Additionally, if the performing agent does not have the ability it will cause a NullPointerException. Let's throw an IllegalActionException if the performing agent doesn't have the necessary ability.